source/sayAll.py

```diff
@@ -1,273 +1,289 @@
-			cb,
-			EndUtteranceCommand()
-		])
-
-	def stop(self):
-		if not self.reader:
-			return
-		self.reader = None
-		self.trigger.exit()
-		self.trigger = None
-
-	def __del__(self):
-		self.stop()
-
-class SayAllProfileTrigger(config.ProfileTrigger):
-	"""A configuration profile trigger for when say all is in progress.
-	"""
-	spec = "sayAll"
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2006-2021 NV Access Limited
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+from typing import Literal
+import weakref
+import garbageHandler
+import speech
+from logHandler import log
+import config
+import controlTypes
+import api
+import textInfos
+import queueHandler
+import winKernel
+
+from speech.commands import CallbackCommand, EndUtteranceCommand
+
+CURSOR_CARET = 0
+CURSOR_REVIEW = 1
+CURSOR = Literal[CURSOR_REVIEW, CURSOR_CARET]
+
+
+class _SayAllHandler:
+	def __init__(self):
+		self.lastSayAllMode = None
+		self._speechWithoutPausesInstance = None
+		#: The active say all manager.
+		#: This is a weakref because the manager should be allowed to die once say all is complete.
+		self._getActiveSayAll = lambda: None  # noqa: Return None when called like a dead weakref.
+
+	@property
+	def speechWithoutPausesInstance(self) -> 'speech.SpeechWithoutPauses':
+		if self._speechWithoutPausesInstance is None:
+			log.error('attempting to use sayAll without calling initialize first')
+			self.initializeSpeechWithoutPauses()  # TODO: or raise Error?
+			# raise AttributeError('call initializeSpeechWithoutPauses first before using sayAllHandler')
+		return self._speechWithoutPausesInstance
+
+	def initializeSpeechWithoutPauses(self):
+		'''
+		Initializes speech.SpeechWithoutPauses, as this cannot always be done when `SayAllHandler` is being 
+		imported without causing circular dependency issues.
+		'''
+		self._speechWithoutPausesInstance = speech.SpeechWithoutPauses(speakFunc=speech.speak)
+
+	def stop(self):
+		'''
+		Stops any active objects reader and resets the SayAllHandler's speech.SpeechWithoutPauses instance
+		'''
+		active = self._getActiveSayAll()
+		if active:
+			active.stop()
+		self.speechWithoutPausesInstance.reset()
+
+	def isRunning(self):
+		"""Determine whether say all is currently running.
+		@return: C{True} if say all is currently running, C{False} if not.
+		@rtype: bool
+		"""
+		return bool(self._getActiveSayAll())
+
+	def readObjects(self, obj: 'NVDAObjects.NVDAObject'):
+		reader = _ObjectsReader(obj)
+		self._getActiveSayAll = weakref.ref(reader)
+		reader.next()
+
+	def readText(self, cursor: CURSOR):
+		self.lastSayAllMode = cursor
+		try:
+			reader = _TextReader(cursor)
+		except NotImplementedError:
+			log.debugWarning("Unable to make reader", exc_info=True)
+			return
+		self._getActiveSayAll = weakref.ref(reader)
+		reader.nextLine()
+
+
+SayAllHandler = _SayAllHandler()
+
```

Comment:
should I make all the methods of `_SayAllHandler` static instead of creating an instance?